### PR TITLE
Adding Roo::Base#close to delete any temp directories

### LINKED ISSUE #210: Add a close method to close any resources, including removing temporary files
Pull request coming shortly

### PR DESCRIPTION
**Pull request by afn**
Fixes #210

### CODE REVIEW

File: lib/roo/base.rb [hunk 1]
```diff
@@ -32,6 +32,14 @@ def initialize(filename, options = {}, _file_warning = :error, _tmpdir = nil)
     @last_column = {}
 
     @header_line = 1
+  rescue => e
```
Comment (PikachuEXE) [inline lib/roo/base.rb]: Should it be `rescue` or `finally`?
Using `rescue` means this only "closed" when exception is caught

**Comment by stevendaniels** [comment]
I think Ruby uses `ensure` whereas Java uses `finally`. One issue with `ensure` is it won't run if a thread is killed. I think using `ensure` is an good approach.

File: lib/roo/base.rb (hunk 1 above)
Comment (stevendaniels) [inline lib/roo/base.rb]: Use `ensure` instead of `rescue`.

Comment (afn) [inline lib/roo/base.rb]: No, it should be `rescue`.

You don't want the connection closed immediately after `initialize` returns.  You only want `initialize` to close the connection if it **doesn't** return.

Comment (stevendaniels) [inline lib/roo/base.rb]: @afn You're right! Thanks for clearing that up.

Comment (PikachuEXE) [inline lib/roo/base.rb]: Would be much clearer if a comment is left on that line...

Comment (simonoff) [inline lib/roo/base.rb]: +1 Comments need to be

**Comment by simonoff** [comment]
Need to be rebased with latest master

**Comment by afn** [comment]
Rebased.

**Comment by simonoff** [comment]
@afn I suppose what rebase was with issues. Tests failed.

**Comment by afn** [comment]
Whoops, sorry about that.  Should be fixed now :)

File: test/test_roo.rb [hunk 2]
```diff
@@ -64,7 +64,7 @@ def with_each_spreadsheet(options)
         yield Roo::Spreadsheet.open(File.join(TESTDIR,
           fixture_filename(options[:name], format)))
       rescue => e
-        raise e, "#{e.message} for #{format}", e.backtrace
+        options[:ignore_errors] or raise e, "#{e.message} for #{format}", e.backtrace
```
Comment (simonoff) [inline test/test_roo.rb]: I don't like this.
Better next
```ruby
raise e, "#{e.message} for #{format}", e.backtrace unless options[:ignore_errors]
```

Comment (afn) [inline test/test_roo.rb]: Changed.

File: test/test_roo.rb [hunk 3]
```diff
@@ -2071,4 +2071,18 @@ def test_open_stream
     xlsx = Roo::Excelx.new(stream)
     assert_equal ["Tabelle1","Name of Sheet 2","Sheet3","Sheet4","Sheet5"], xlsx.sheets
   end
+
+  def test_close
+    with_each_spreadsheet(:name=>'numbers1') do |oo|
+      next unless tempdir = oo.instance_variable_get('@tmpdir')
```
Comment (simonoff) [inline test/test_roo.rb]: Better
```ruby
next unless (tempdir = oo.instance_variable_get('@tmpdir'))
```

File: lib/roo/base.rb [hunk 4]
```diff
@@ -521,7 +533,9 @@ def make_tmpdir(prefix = nil, root = nil, &block)
     else
       TEMP_PREFIX
     end
-    Dir.mktmpdir(prefix, root || ENV['ROO_TMP'], &block)
+    Dir.mktmpdir(prefix, root || ENV['ROO_TMP'], &block).tap do |result|
+      block_given? or track_tmpdir! result
```
Comment (simonoff) [inline lib/roo/base.rb]: Please avoid using `or`

File: lib/roo/base.rb [hunk 5]
```diff
@@ -435,6 +443,10 @@ def is_stream?(filename_or_stream)
 
   private
 
+  def track_tmpdir! tmpdir
```
Comment (simonoff) [inline lib/roo/base.rb]: Please make 
```ruby
def track_tmpdir!(tmpdir)
```

File: lib/roo/base.rb [hunk 6]
```diff
@@ -32,6 +32,14 @@ def initialize(filename, options = {}, _file_warning = :error, _tmpdir = nil)
     @last_column = {}
 
     @header_line = 1
+  rescue => e # clean up any temp files, but only if an error was raised
+    close
+    raise e
+  end
+
+  def close
+    @tmpdirs.nil? or @tmpdirs.each { |dir| FileUtils.remove_entry dir }
```
Comment (simonoff) [inline lib/roo/base.rb]: Same, please avoid using `or`

Comment (simonoff) [inline lib/roo/base.rb]: There will be better to have next:
```ruby
def close
   return nil unless @tmpdirs
   @tmpdirs.each { |dir| FileUtils.remove_entry dir }
  nil
end
```

**Comment by afn** [comment]
@simonoff: Great — thanks for the detailed feedback! Since these comments are all stylistic in nature, I'll let you take the lead and make the necessary changes to ensure that the style is consistent with the rest of the codebase.

**Pull request merged**